drivers/timer/nrf_rtc_timalarm.c

```diff
@@ -0,0 +1,459 @@
+	 * passed (causing immediate callback) if uncounted ticks have
+	 * occurred.
+	 *
+	 * A positive tick schedules for the requested number of ticks
+	 * after the last announced tick.  This too may be in the
+	 * past.
+	 */
+
+	/* Explain why we need to do this. */
+	(void)k_alarm_cancel(&tick_alarm);
+
+	if (ticks == K_FOREVER) {
+		/* "no future timer interrupts are expected or
+		 * required"
+		 *
+		 * At this point we have no obligation to maintain the
+		 * tick clock; consequently in tick calculation code
+		 * we are entitled to act as though last_tick_sc is
+		 * within a half 32-bit span of the current time, an
+		 * invariant maintained by MAX_TICKS.
+		 */
+		return;
+	}
+
+	u32_t nts;
+	k_spinlock_key_t key = k_spin_lock(&lock);
+	u32_t ntd = (sysclock_get_32() - last_tick_sc) / SC_PER_TICK;
+
+	if (ntd == 0) {
+		/* Can't announce zero ticks */
+		ntd = 1;
+	}
+	if (ticks > 0) {
+		/* Kernel wants a timeout with a positive offset
+		 * relative to the last announced tick.
+		 */
+		ntd = MAX(ntd, MIN(ticks, MAX_TICKS));
+	}
+	next_tick_delta = ntd;
+	next_tick_sc = last_tick_sc + next_tick_delta * SC_PER_TICK;
+	nts = next_tick_sc;
+	k_spin_unlock(&lock, key);
+
+	(void)k_alarm_schedule(&tick_alarm, nts, 0);
+#endif
+}
+
+u32_t z_clock_elapsed(void)
+{
+	u32_t rv = 0;
+#ifdef CONFIG_TICKLESS_KERNEL
+	k_spinlock_key_t key = k_spin_lock(&lock);
+
+	rv = (sysclock_get_32() - last_tick_sc) / SC_PER_TICK;
+
+	k_spin_unlock(&lock, key);
+#endif
+	return rv;
+}
+
+/* Implement z_alarm_update_deadline. */
+static void alarm_update_deadline_(u32_t now)
+{
+	u32_t compare = now;
+	u32_t deadline = 0;
+	int rc = k_alarm_next_deadline_(&deadline);
+
+	if (rc < 0) {
+		/* No unscheduled alarms, use maximum delay */
+		compare += COUNTER_MAX_DELTA;
+	} else if (rc > 0) {
+		/* Next event at deadline.  If that's now or in the
+		 * past target a compare of now; otherwise queue an
+		 * event at that deadline or the maximum delta from
+		 * now whichever's sooner.
+		 */
+		u32_t delay = deadline - now;
+
+		if ((s32_t)delay > 0) {
+			if (delay < COUNTER_MAX_DELTA) {
+				compare += delay;
+			} else {
+				compare += COUNTER_MAX_DELTA;
+			}
+		} else {
+			/* Alarm is (past) due, fire ASAP */
+		}
+	} else {
+		/* Something is ready, fire ASAP */
+	}
+
+	/* compare is at most COUNTER_MAX_DELTA past now.  The RTC
+	 * counter should not have advanced more than
+	 * COUNTER_MAX_DELTA - 2 ticks past now.
+	 *
+	 * If the next event is already due and we haven't yet cleared
+	 * the last event, leave it enabled so we'll re-enter the
+	 * FLIH.
+	 */
+	if ((compare == now)
+	    && RTC->EVENTS_COMPARE[0]) {
+		return;
+	}
+
+	/* Make sure compare is at least COUNTER_MIN_DELTA past now,
+	 * then clear the COMPARE event and set the compare value.
+	 */
+	if (COUNTER_MIN_DELTA > (compare - now)) {
+		compare = now + COUNTER_MIN_DELTA;
+	}
+	RTC->EVENTS_COMPARE[0] = 0;
+	RTC->CC[0] = compare;
+}
+
+void z_alarm_update_deadline(void)
```

Comment:
I think it was used elsewhere in earlier versions.  I also think it's clearer as a separate function than in the body of a conditional, but I do think it should be made `ALWAYS_INLINE` since there's a single call site and a stupid compiler might fail to do the right thing.